File: practice/implementation/the_grid_search/the_grid_search.hs
import Control.Applicative
import Control.Monad
import System.IO
import qualified Data.Matrix as Matrix
import Data.List

getMultipleLines :: Int -> IO [String]

getMultipleLines n
    | n <= 0 = return []
    | otherwise = do          
        x <- getLine         
        xs <- getMultipleLines (n-1)    
        let ret = (x:xs)    
        return ret

contains mgrid mpattern rg cg rp cp = do
    any (==True) [Matrix.submatrix r (r + rp - 1) c (c + cp - 1) mgrid ==  mpattern | r <- [1..(rg - rp + 1)], c <- [1..(cg - cp + 1)]]

main :: IO ()
main = do
    t <- readLn :: IO Int
    forM_ [1..t] $ \_  -> do
        g_temp <- getLine
        let g_t = words g_temp
            rg = read $ g_t!!0 :: Int
            cg = read $ g_t!!1 :: Int
        grid <- getMultipleLines rg
        
        r_temp <- getLine
        let r_t = words r_temp
            rp = read $ r_t!!0 :: Int
            cp = read $ r_t!!1 :: Int
        pattern <- getMultipleLines rp    
        let ans = contains (Matrix.fromLists grid) (Matrix.fromLists pattern) rg cg rp cp
        if ans
            then putStrLn "YES"
            else putStrLn "NO"

File: practice/implementation/the_grid_search/the_grid_search.py
#!/bin/python

def contains(grid, pattern, row, col):
    for ROW in range(len(pattern)):
        for COL in range(len(pattern[0])):
            x = ROW + row
            y = COL + col
            if (x >= len(grid)) or (y >= len(grid[0])):
                return False
            if grid[x][y] != pattern[ROW][COL]:
                return False
    return True

t = int(raw_input().strip())
for _ in range(t):
    R, C = map(int, raw_input().strip().split())
    grid = []
    for __ in range(R):
        grid.append(str(raw_input().strip()))

    r, c = map(int, raw_input().strip().split())
    pattern = []
    for __ in range(r):
        pattern.append(str(raw_input().strip()))
        
    ans = False
    for ROW in range(R):
        for COL in range(C):
            if contains(grid, pattern, ROW, COL):
                ans = True
                break
        if ans:
            break
    if ans:
        print "YES"
    else:
        print "NO"

